### PR TITLE
Add support for BitBoxBase in the Python scripts

### PR DESCRIPTION
**Pull request by conte91**
Still very WIP.

### CODE REVIEW

**Comment by conte91** [comment]
> This is very hard to review as it is. Every commit should be "an atomic functional change", pass the linters and work. I see temporary code added and later removed.
> 

Ouch, the first commit is indeed coming from the hacks we used to get the very first version working. I'll need to squash it.

**Comment by conte91** [comment]
> The only bug I found is in `load_firmware` that looks at `argv` even though `argparse` is used.
> 

This is a leftover as well, will fix.

File: py/bitbox02/bitboxbase/bitboxbase.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    # pylint: disable=too-many-public-methods
+
+    def __init__(
+        self,
+        device: communication.TransportLayer,
+        device_info: DeviceInfo,
+        show_pairing_callback: Callable[[str], None],
+        attestation_check_callback: Optional[Callable[[bool], None]] = None,
+    ):
+        communication.BitBoxAPIExchanger.__init__(
```
Comment (benma) [inline py/bitbox02/bitboxbase/bitboxbase.py]: same

File: py/bitbox02/bitbox02/bootloader.py [hunk 2]
```diff
@@ -66,6 +66,7 @@ def __init__(self, transport: TransportLayer, device_info: DeviceInfo):
         self.expected_magic = {
             "bb02-bootloader": SIGDATA_MAGIC_STANDARD,
             "bb02btc-bootloader": SIGDATA_MAGIC_BTCONLY,
+            "bb02-base": SIGDATA_MAGIC_STANDARD,
```
Comment (benma) [inline py/bitbox02/bitbox02/bootloader.py]: please add a `SIGDATA_MAGIC_BITBOXBASE_STANDARD = 0xab6bd345` (this is the new magic) and use it here.

File: py/load_firmware.py [hunk 3]
```diff
@@ -62,28 +66,13 @@ def _show_pairing(code: str) -> None:
...
-            "should be signed."
-        )
-        return 1
-
-    filename = sys.argv[1]
-    if not debug and ".signed.bin" not in filename:
-        eprint("Expecting firmware to end with '.signed.bin'")
-        return 1
-
+def find_and_open_usb_bitbox() -> (dict, TransportLayer):
```
Comment (benma) [inline py/load_firmware.py]: `bitbox02` :P

Comment (benma) [inline py/load_firmware.py]: maybe underscore to mark as private

File: py/load_firmware.py [hunk 4]
```diff
@@ -92,19 +81,107 @@ def main() -> int:
...
+
+
+def try_usart_bootloader_connection(serial_port, bootloader_device) -> str:
+    transport = usart.U2FUsart(serial_port)
+    try:
+        bootloader_attempt = Bootloader(transport, bootloader_device)
+        bootloader_attempt.versions()
+        success = "Connected"
+    except usart.U2FUsartErrorResponse as e:
+        if (
```
Comment (benma) [inline py/load_firmware.py]: no parens in python ;)

File: py/load_firmware.py [hunk 5]
```diff
@@ -92,19 +81,107 @@ def main() -> int:
...
+    # The bootloader wasn't valid, try to connect to the firmware instead.
+    print("BitBox bootloader not available.")
+    print("Trying to connect to BitBox firmware instead...")
+
+    def _show_pairing(code: str) -> None:
+        print("Accept the pairing request on your device...")
+
+    try:
+        transport = usart.U2FUsart(serial_port)
+        bitbox_attempt = BitBox02(transport, bootloader_device, show_pairing_callback=_show_pairing)
```
Comment (benma) [inline py/load_firmware.py]: BitBoxBase

File: py/load_firmware.py [hunk 6]
```diff
@@ -92,19 +81,107 @@ def main() -> int:
...
+            sys.exit(1)
 
     pprint.pprint(bootloader_device)
 
     hid_dev = hid.device()
     hid_dev.open_path(bootloader_device["path"])
+    return bootloader_device, u2fhid.U2FHid(hid_dev)
+
+
+def try_usart_bootloader_connection(serial_port, bootloader_device) -> str:
```
Comment (benma) [inline py/load_firmware.py]: return `enum`, not string.

File: py/bitbox02/bitbox02/bootloader.py [hunk 7]
```diff
@@ -60,19 +61,17 @@ class Bootloader:
...
             "bb02-bootloader": SIGDATA_MAGIC_STANDARD,
             "bb02btc-bootloader": SIGDATA_MAGIC_BTCONLY,
         }.get(device_info["product_string"])
         assert self.expected_magic
 
     def _query(self, msg: bytes) -> bytes:
-        cid = u2fhid.generate_cid()
-        u2fhid.write(self.device, msg, BOOTLOADER_CMD, cid)
-        response = bytes(u2fhid.read(self.device, BOOTLOADER_CMD, cid))
+        cid = self.device.generate_cid()
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/bitbox02/bitbox02/bootloader.py]: The `u2f.bitbox02` module needs explicit CID generation (as it needs to use the same CID throughout multiple transactions)

File: py/send_message.py [hunk 8]
```diff
@@ -509,7 +516,10 @@ def attestation_check(result: bool) -> None:
             else:
                 print("Device attestation FAILED")
 
-        device = bitbox02.BitBox02(
+        hid_device = hid.device()
+        hid_device.open_path(bitbox["path"])
+        bitbox_connection = bitbox02.BitBox02(
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/send_message.py]: Well, it's a application-layer connection to the device's API. `bitbox02_api_connection` then?

File: py/bitbox02/bitbox02/bootloader.py [hunk 9]
```diff
@@ -60,19 +61,17 @@ class Bootloader:
     One instance of a BitBox02 Bootloader, exposing the bootloader API.
     """
 
-    def __init__(self, device_info: DeviceInfo):
-        self.device = hid.device()
-        self.device.open_path(device_info["path"])
+    def __init__(self, transport: TransportLayer, device_info: DeviceInfo):
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/bitbox02/bitbox02/bootloader.py]: Fixed.

File: py/bitbox02/communication/__init__.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Abstraction for the transport layer used for transmitting U2F messages."""
+
+from .communication import *
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/bitbox02/communication/__init__.py]: Fixed.

File: py/bitbox02/communication/u2fhid/u2fhid.py [hunk 11]
```diff
@@ -41,128 +42,122 @@
...
-            idx += len(buf) - 5
-        if reply_cid != cid:
-            raise Exception(f"- USB channel ID mismatch {reply_cid:x} != {cid:x}")
-        if reply_cmd != cmd:
-            raise Exception(f"- USB command mismatch {reply_cmd:x} != {cmd:x}")
-        return data[:data_len]
-    raise Exception("Did not read anything after %d seconds" % timeout)
+# U2F-over-HID transport layer
+class U2FHid(TransportLayer):
+    def __init__(self, hid_device: PhysicalLayer):
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/bitbox02/communication/u2fhid/u2fhid.py]: Fixed.

File: py/load_firmware.py [hunk 12]
```diff
@@ -94,7 +99,10 @@ def main() -> int:
 
     pprint.pprint(bootloader_device)
 
-    bootloader = Bootloader(bootloader_device)
+    hid_dev = hid.device()
```
[the remark this answers is not in the document]
Comment (conte91) [reply py/load_firmware.py]: Fixed.

File: py/load_firmware.py (hunk 4 above)
Comment (conte91) [reply py/load_firmware.py]: Fixed.

File: py/load_firmware.py (hunk 6 above)
Comment (conte91) [reply py/load_firmware.py]: Fixed.

File: py/load_firmware.py (hunk 5 above)
Comment (conte91) [reply py/load_firmware.py]: Fixed.

File: py/load_firmware.py (hunk 3 above)
Comment (conte91) [reply py/load_firmware.py]: Fixed.

Comment (conte91) [reply py/load_firmware.py]: Fixed.

File: py/bitbox02/communication/u2fhid/u2fhid.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+            The read message combined from the u2fhid packets
+        Throws:
+            ValueError: In case any value is out of range
+            Exception: In case of USB communication issues
+        """
+        if endpoint < 0 or endpoint > 0xFF:
+            raise ValueError("Endpoint/U2F command is out of range '0 < endpoint <= 0xFF'")
+        if cid < 0 or cid > 0xFFFFFFFF:
+            raise ValueError("Channel id is out of range '0 < cid <= 0xFFFFFFFF'")
+        timeout_ms = 5000
```
Comment (benma) [inline py/bitbox02/communication/u2fhid/u2fhid.py]: this timeout is too short, can't even unlock the device. the previous value was 5000000s (basically infinite). please restore it to this value.

Comment (benma) [reply py/bitbox02/communication/u2fhid/u2fhid.py]: maybe hid device read can take an argument for no timeout (0, None, negative, ...).

Comment (benma) [reply py/bitbox02/communication/u2fhid/u2fhid.py]: are there any other functional changes in the the non-usart code? I expected that the regular bitbox02/bootloader/usb/hid not to change at all, functionally.

Comment (conte91) [reply py/bitbox02/communication/u2fhid/u2fhid.py]: > this timeout is too short, can't even unlock the device. the previous value was 5000000s (basically infinite). please restore it to this value.

Was 5000s before - This is a typo due to missing the fact that the first argument was in seconds, and the `hid` argument is in ms instead. Will revert.

> are there any other functional changes in the the non-usart code? I expected that the regular bitbox02/bootloader/usb/hid not to change at all, functionally.

It should be unchanged, indeed. A quick diff (with whitespace ignored) shows no difference (apart from the 
timeout typo and changes needed for making the class compatible with the `TransportLayer` class).

Also there is the (pretty evident) rename of the generated files from `bitbox02/generated` to `communication/generated`.

Comment (benma) [reply py/bitbox02/communication/u2fhid/u2fhid.py]: ok np ;) 

there was a hacky `timeout = timeout * 1000` before that dropped: https://github.com/digitalbitbox/bitbox02-firmware/pull/203/files#diff-420ffb3dcc6af5c0c0d4f7254829149dL146

it was seconds, not milliseconds, all along.

no one to blame but myself for quickly putting this in to fix the timeout issue right before release :D

Comment (benma) [reply py/bitbox02/communication/u2fhid/u2fhid.py]: anyway, with 5s most of the blocking workflows, like reading the mnemonic, will fail.

Comment (conte91) [reply py/bitbox02/communication/u2fhid/u2fhid.py]: Fixed

**Pull request merged**